nav2_smac_planner/src/smac_planner_2d.cpp

```diff
@@ -80,6 +80,9 @@ void SmacPlanner2D::configure(
   nav2_util::declare_parameter_if_not_declared(
     node, name + ".max_on_approach_iterations", rclcpp::ParameterValue(1000));
   node->get_parameter(name + ".max_on_approach_iterations", _max_on_approach_iterations);
+  nav2_util::declare_parameter_if_not_declared(
+    node, name + ".ignore_goal_orientation", rclcpp::ParameterValue(false));
+  node->get_parameter(name + ".ignore_goal_orientation", _ignore_goal_orientation);
```

Comment:
No need for a parameter here, this should always just be done. There's no situation I can think of where this isn't the right thing to do.

Reply:
With this parameter set to false (default), the goal pose is respected, which is the classical behavior of NavFn and other planner, i.e last path pose orientation = goal pose orientation
When it is set to true, the last pose has the same orientation as the n - 1 pose, which is a behavior that I sometimes need (i.e. send the robot to a `position` without orientation constraints, as opposite to send the robot to a `pose` with orientation constraints. And apparently I am not the only one, see https://github.com/ros-planning/navigation2/issues/2478#issuecomment-891834153, though it is for the hybrid version.

Reply:
Weird, I'm not regularly getting notifications about this ticket, sorry about that. 

For Hybrid-A*, there will not (and impossible to be) a version of that planner without orientations, it is by nature kinematically feasible which means you need a full kinematic state to hit. That person is misinformed about path planning. 

> With this parameter set to false (default), the goal pose is respected, which is the classical behavior of NavFn and other planner, i.e last path pose orientation = goal pose orientation

I think that's the right behavior for 2D always. If you don't want orientation constraints, why are you not just setting your goal checker plugin to have a tolerance of 2*PI? Any orientation is valid then. We should not have the path planner just "ignore" valid state information. I know for the 2D case its pretty silly because there's no orientations in the algorithm, but if you request `10 deg`, even if you don't really mean it, the planner should just take that at face value to return a goal at `10 deg`. 

I think the place to ignore that information is the controller server. The goal checker can be set to 2*PI to make it just care about the position and not at all about the orientation. That would result in the same behavior and much more robust to changes / path tracking errors (e.g. if you set to n-1 but then approached at a slightly different angle using a modeling based local planner, then it would try to rotate to n-1's heading)

Reply:
> If you don't want orientation constraints, why are you not just setting your goal checker plugin to have a tolerance of 2*PI?

Well, because I use a local planner/controller which looks way ahead of its current pose and is allowed to go away from the global path in order to time optimize its trajectory.  Hence even with a angle tolerance of 2*PI, the final pose orientation will have an effect on the trajectory computed by my local planner/controller. But better explained with gifs:

1. Goal orientation pointed left of screen and hence equal to the final global path pose orientation, 2*PI angle tolerance, 0.02m distance tolerance
![without_ignore_orientation](https://user-images.githubusercontent.com/15727892/128945952-ad16ab9c-6f44-4388-90cd-392445b7cd96.gif)

2. Goal orientation pointed left of screen BUT ignore_goal_orientation set to true so final global path pose orientation equal n-1 pose orientation , 2*PI angle tolerance, 0.02m distance tolerance
![with_ignore_orientation](https://user-images.githubusercontent.com/15727892/128946354-bd91096e-6f31-46d7-8326-ef3a1ebecff7.gif)

> We should not have the path planner just "ignore" valid state information

I agree that this should definitely not be the default behavior but there is the use case of wanting to go to a `Position` instead of a `Pose`

> I think the place to ignore that information is the controller server. The goal checker can be set to 2*PI to make it just care about the position and not at all about the orientation. That would result in the same behavior and much more robust to changes / path tracking errors (e.g. if you set to n-1 but then approached at a slightly different angle using a modeling based local planner, then it would try to rotate to n-1's heading). So the real solution is in the goal checker plugin, which if you do, makes this moot.

I am not sure what is the right place for it, but as shown above, the goal checker is not the way to solve this.
I believe there is an use case for it (or at least I have it) and to me it will nicely fit in the 2D planner as an optional behavior (as it is an orientation ignorant algorithm anyway). However if you feel that it is better to leave it out of the planner, the cost for me to put that end pose orientation modification somewhere else, like in my custom planner server, is small, and I will remove it from this PR.

Reply:
That's a fair situation. Mhm. I do wonder what is best here. Thematically, it seems wrong for us to just "ignore" the goal requested orientation, and if we did, it would seems even a weirder behavior to default it to the approach angle rather than just being `0.0` as the default value of a `double`. How does the controller handle just a default `0.0`?

But at the same time, I can appreciate the situation. The other half of my brain then says "ok, then let's solve this for the general case" which would then be somewhere in the planner server so that it could apply to any planner and not specifically just the Smac planner (and only the 2D version of that, no less). 

Neither of these situations are the options you're describing as desirable though. I agree for your predictive controller you'd need to manipulate the planner's approach angle as well as the goal checker for task termination. 

So ideally, I'd want it in the planner server - which doesn't sound like it would help you much if you have your own planner server. I'd also want to mess with the naming to be more clear to what it is doing since it's not really "ignoring" it as much as "setting" it to the approach angle. 

---

Just curiosity, is that TEB or something else? Why the separate planner server?

Reply:
> That's a fair situation. Mhm. I do wonder what is best here. Thematically, it seems wrong for us to just "ignore" the goal requested orientation, and if we did, it would seems even a weirder behavior to default it to the approach angle rather than just being `0.0` as the default value of a `double`. How does the controller handle just a default `0.0`?

The controller treats it as a 0 deg orientation request

> But at the same time, I can appreciate the situation. The other half of my brain then says "ok, then let's solve this for the general case" which would then be somewhere in the planner server so that it could apply to any planner and not specifically just the Smac planner (and only the 2D version of that, no less).

The thing is, for "smart" planners, like the hybrid, where the goal pose orientation has an influence on many poses before the last, the trick of copying the n-1 orientation is not working. Putting such a parameter in the planner server would imply that all planners can make use of it, which for some planner is complex (or maybe impossible). Or do you see a way of having it in the planner server that will not confuse users in case the planner they use is not implementing this feature ? 
I like the idea of putting it in the planner itself because it is immediately clear if this feature is supported: the parameter is present, it is, the parameter is absent, it is not. But I don't have a strong opinion on this.


> Neither of these situations are the options you're describing as desirable though. I agree for your predictive controller you'd need to manipulate the planner's approach angle as well as the goal checker for task termination.
> 
> So ideally, I'd want it in the planner server - which doesn't sound like it would help you much if you have your own planner server. I'd also want to mess with the naming to be more clear to what it is doing since it's not really "ignoring" it as much as "setting" it to the approach angle. This topic will probably come up again, so I'm not entirely sure what the right answer is, maybe add that parameter to this and NavFn / Theta* so all the 2D planners have this option? That seems less.. great than having the planner server take it on, but what if you have multiple planners and only want this behavior with 1? There are other parameters like `unknown` and `tolerance` in each multiple planners implement, so maybe this could just be another?

Now that you mention it, yes, that could be another parameter like `unknown` and `tolerance` and that's an argument for putting it in the planners.
About the name we can brainstorm. But I would like it to convey the idea that we tell the planner to not use the last goal orientation as a constraint to compute its path, which make it more general than copying the approach angle. compute_without_goal_orientation_constraint ? 2d_only_goal ? xy_only_goal ? 

> I know that's not a very clear "do this" answer. The only thing I can clearly state is changing the name. Otherwise, if it goes into Smac2D as a param, it should also go in NavFn / Theta* to match. If it goes into planner server, then no need. What do you think?

I prefer it in the planner, and I can add it the NavFn, but I will let you take the decision.

> Just curiosity, is that TEB or something else? Why the separate planner server?

Yes, TEB based. Not entirely separate planner server, it still inherits from nav2_planner::PlannerServer and we have a couple of overload mainly for dealing with computing path transparently when the robot is docked, maintaining multiple costmaps in parallel with different options and custom feedback/result messages

Reply:
`use_final_approach_orientation` would be my naming suggestion.

BTW any parameter changes needs a complementary update to navigation.ros.org for any packages changed. Also adding to the migration guide this new parameter and what it does (those 2 gifs would be helpful illustrations).

Lets do this, lets add this parameter to all of the 2D planners. If true, does the approach angle. If false, sets it to the requested goal orientation. So that's Navfn, Smac2D, and Theta*.

Reply:
A small test in Smac2D and Theta* at minimum would be nice to make sure this is working when set

Reply:
All right, I will try to attend asap

Reply:
Ok, added to NavFn and Theta*. Theta* had also the same end pose orientation always set to 0 issue as Smac2D, fixed.
Tests where added to test_planner_pluggins.cpp and I changed the name from `test_planner_plugin_failures` to `test_planner_plugins`. It was easy to do with the **nav2_system_tests::NavFnPlannerTester** class, which btw can be used to test other planner than NavFn, so maybe we should change its name ?
Let me know if you want another iteration, and if that's good, I will prepare the documentation PR